### PR TITLE
Create empty fbc index image

### PR DESCRIPTION
**Pull request by chandwanitulsi**
In order to create an empty FBC index image,
a new parameter 'output_fbc' is added to create-empty-index
endpoint. When this parameter is specified an empty
index image created will be created using the from_index
image.
Refers to CLOUDDST-8773

### CODE REVIEW

File: iib/workers/tasks/opm_operations.py [hunk 1]
```diff
@@ -382,3 +383,98 @@ def opm_registry_add_fbc(
...
+    """
+    # if from_index provided is FBC, get the hidden index db location
+    if is_image_fbc(from_index_resolved):
+        log.debug(f'{from_index} provided is FBC index image')
+        index_db_path = get_hidden_index_database(from_index=from_index, base_dir=temp_dir)
+    # if the from_index is sqllite based, get the default index db location
+    else:
+        from iib.workers.tasks.build import _get_index_database
+
+        log.debug(f'{from_index} provided is sqllite index image')
```
Comment (lipoja) [inline iib/workers/tasks/opm_operations.py:465]: Could you please use % string formatting with logging. It is optimized for that.
`log.debug('%s provided is sqllite index image', from_index)`

File: iib/workers/tasks/opm_operations.py [hunk 2]
```diff
@@ -382,3 +383,98 @@ def opm_registry_add_fbc(
...
+    :param str from_index: the pull specification of the container image containing the index that
+        the index image build will be based from.
+    :param str binary_image: the pull specification of the container image where the opm binary
+        gets copied from. This should point to a digest or stable tag.
+    :param list operators: a list of strings representing the packages of the operators to be
+        removed from the output index image.
+    """
+    # if from_index provided is FBC, get the hidden index db location
+    if is_image_fbc(from_index_resolved):
+        log.debug(f'{from_index} provided is FBC index image')
```
Comment (lipoja) [inline iib/workers/tasks/opm_operations.py:459]: See comment below.

**Review by lipoja — CHANGES_REQUESTED** [review]

File: iib/web/api_v1.py [hunk 3]
```diff
@@ -895,6 +895,7 @@ def create_empty_index():
     args = [
         payload['from_index'],
         request.id,
+        payload.get('fbc'),
```
[the remark this answers is not in the document]
Comment (chandwanitulsi) [reply iib/web/api_v1.py:898]: I wouldn't mind changing it to 'create_fbc', won't create_fbc be redundant in create-empty-index api

File: iib/web/static/api_v1.yaml [hunk 4]
```diff
@@ -1143,6 +1143,9 @@ components:
           items:
             type: string
           example: ["v4.5-10-08-2021"]
+        fbc:
+          type: boolean
+          description: Specifies that output index image should be FBC
```
[the remark this answers is not in the document]
Comment (chandwanitulsi) [reply iib/web/static/api_v1.yaml:1148]: makes sense :+1:

File: iib/workers/tasks/build_create_empty_index.py [hunk 5]
```diff
@@ -86,16 +87,30 @@ def handle_create_empty_index_request(
...
         set_request_state(request_id, 'in_progress', 'Checking operators present in index image')
 
         operators = _get_present_operators(from_index_resolved, temp_dir)
-        set_request_state(request_id, 'in_progress', 'Removing operators from index image')
 
-        _opm_index_rm(temp_dir, operators, prebuild_info['binary_image'], from_index_resolved)
+        # if fbc parameter is true, create an empty FBC index image
+        # else create sqllite index image
+        if fbc:
+            log.debug(f'Creating empty FBC from {from_index}')
```
Comment (chandwanitulsi) [inline iib/workers/tasks/build_create_empty_index.py:97]: I forgot to change this, let me change in next patch

File: iib/web/models.py [hunk 6]
```diff
@@ -1543,6 +1544,7 @@ class RequestCreateEmptyIndex(Request, RequestIndexImageMixin):
         'polymorphic_identity': RequestTypeMapping.__members__['create_empty_index'].value
     }
     build_tags = None
+    fbc = None
```
Comment (JAVGan) [inline iib/web/models.py:1548]: Since it will receive a boolean wouldn't it be better to assign `false` as default?

Or do we need to have a differentiation between `unset` and `bool`?

File: iib/workers/tasks/build.py [hunk 7]
```diff
@@ -929,17 +934,22 @@ def handle_rm_request(
...
-                err_msg = 'File-Based catalog image type is not supported yet.'
-                log.error(err_msg)
-                raise IIBError(err_msg)
+                log.info("Processing File-Based Catalog image")
+                opm_registry_rm_fbc(
+                    base_dir=temp_dir,
+                    from_index=from_index_resolved,
+                    operators=operators,
+                    binary_image=prebuild_info['binary_image'],
+                )
```
Comment (JAVGan) [inline iib/workers/tasks/build.py:943]: Is this related to `create-empty-index`?

Because it's touching `handle_rm_request` and I would like to know whether is a requirement for `cei` it or not.

File: iib/web/models.py (hunk 6 above)
Comment (chandwanitulsi) [reply iib/web/models.py:1548]: yeah, I can change that

File: iib/workers/tasks/build.py (hunk 7 above)
Comment (chandwanitulsi) [reply iib/workers/tasks/build.py:943]: this is from another commit related to remove operator, once  #[342] is merged and I will rebase this over that

File: iib/workers/tasks/opm_operations.py [hunk 8]
```diff
@@ -382,3 +384,101 @@ def opm_registry_add_fbc(
         binary_image=binary_image,
         dockerfile_name='index.Dockerfile',
     )
+
+
+def _opm_registry_rm(index_db_path, operators, base_dir):
+    """
+    Generate and run the opm command to remove operator package from index db provided.
```
Comment (JAVGan) [inline iib/workers/tasks/opm_operations.py:391]: Same `rm` here

File: iib/workers/tasks/opm_operations.py [hunk 9]
```diff
@@ -382,3 +384,101 @@ def opm_registry_add_fbc(
...
+        '--packages',
+        ','.join(operators),
+    ]
+    run_cmd(cmd, {'cwd': base_dir}, exc_msg='Failed to remove operators from the index image')
+
+
+@retry(exceptions=IIBError, tries=2, logger=log)
+def opm_registry_rm_fbc(base_dir, from_index, operators, binary_image):
+    """
+    Remove operator/s from a File Based Catalog index image.
```
Comment (JAVGan) [inline iib/workers/tasks/opm_operations.py:414]: `rm` here as well

File: tests/test_workers/test_tasks/test_build.py [hunk 10]
```diff
@@ -1019,18 +1019,80 @@ def test_handle_rm_request(
...
+@mock.patch('iib.workers.tasks.build._update_index_image_pull_spec')
+@mock.patch('iib.workers.tasks.build._add_label_to_index')
 @mock.patch('iib.workers.tasks.build.is_image_fbc')
-def test_handle_rm_request_raises(mock_iifbc, mock_runcmd, mock_c):
+@mock.patch('iib.workers.tasks.opm_operations.get_hidden_index_database')
+@mock.patch('iib.workers.tasks.opm_operations.opm_migrate')
+@mock.patch('iib.workers.tasks.opm_operations.opm_generate_dockerfile')
+def test_handle_rm_request_fbc(
+    mock_ogd,
+    mock_om,
```
Comment (JAVGan) [inline tests/test_workers/test_tasks/test_build.py:1040]: And here `rm` test

File: tests/test_workers/test_tasks/test_opm_operations.py [hunk 11]
```diff
@@ -328,3 +328,86 @@ def test_opm_registry_add_fbc(
...
+
+
+@pytest.mark.parametrize('operators', (['abc-operator', 'xyz-operator'], []))
+@mock.patch('iib.workers.tasks.opm_operations.opm_generate_dockerfile')
+@mock.patch('iib.workers.tasks.opm_operations.opm_migrate')
+@mock.patch('iib.workers.tasks.opm_operations._opm_registry_rm')
+@mock.patch('iib.workers.tasks.opm_operations.get_hidden_index_database')
+def test_opm_registry_rm_fbc(
+    mock_ghid, mock_orr, mock_om, mock_ogd, tmpdir, operators,
+):
```
Comment (JAVGan) [inline tests/test_workers/test_tasks/test_opm_operations.py:340]: Another `rm` test

File: tests/test_workers/test_tasks/test_opm_operations.py [hunk 12]
```diff
@@ -328,3 +328,86 @@ def test_opm_registry_add_fbc(
...
+        index_db=index_db_file,
+        binary_image='some:image',
+        dockerfile_name='index.Dockerfile',
+    )
+
+
+@mock.patch('iib.workers.tasks.utils.run_cmd')
+def test_opm_registry_rm(mock_run_cmd):
+    packages = ['abc-operator', 'xyz-operator']
+    opm_operations._opm_registry_rm(
```
Comment (JAVGan) [inline tests/test_workers/test_tasks/test_opm_operations.py:368]: I guess this is the last `rm` test

**Review by JAVGan — COMMENTED** [review]
Everything looks ok but it's mixed with other commit non related to this PR.

File: iib/workers/tasks/build.py (hunk 7 above)
Comment (JAVGan) [reply iib/workers/tasks/build.py:943]: Ok, I've added comments over the places I found that, hope it's useful.

Once it's rebased just let me know then I review it again

File: iib/workers/tasks/build_create_empty_index.py [hunk 13]
```diff
@@ -86,16 +87,30 @@ def handle_create_empty_index_request(
 
     _update_index_image_build_state(request_id, prebuild_info)
     with tempfile.TemporaryDirectory(prefix='iib-') as temp_dir:
-        if is_image_fbc(from_index_resolved):
-            err_msg = 'File-Based catalog image type is not supported yet.'
-            log.error(err_msg)
-            raise IIBError(err_msg)
         set_request_state(request_id, 'in_progress', 'Checking operators present in index image')
 
         operators = _get_present_operators(from_index_resolved, temp_dir)
```
Comment (pbortlov) [inline iib/workers/tasks/build_create_empty_index.py:92]: Q: Is there anyhow taken care of how to get operators when it is FBC image?

File: iib/workers/tasks/opm_operations.py [hunk 14]
```diff
@@ -435,3 +437,48 @@ def opm_registry_rm_fbc(base_dir, from_index, operators, binary_image):
...
+
+def opm_create_empty_fbc(
+    request_id, temp_dir, from_index_resolved, from_index, binary_image, operators
+):
+    """
+    Create empty FBC index image.
+
+    This only produces the index.Dockerfile file and does not build the container image.
+
+    :param str temp_dir: the base directory to generate the database and index.Dockerfile in.
```
Comment (pbortlov) [inline iib/workers/tasks/opm_operations.py:450]: reguest_id is missing here

File: tests/test_workers/test_tasks/test_build_create_empty_index.py [hunk 15]
```diff
@@ -123,9 +124,94 @@ def test_handle_create_empty_index_request_raises(mock_iifbc, mock_c):
     mock_iifbc.return_value = True
     with pytest.raises(IIBError):
         build_create_empty_index.handle_create_empty_index_request(
-            from_index='index-image:latest',
+            from_index='fbc-index-image:latest',
```
Comment (pbortlov) [inline tests/test_workers/test_tasks/test_build_create_empty_index.py:127]: This is not a test for fbc

File: tests/test_workers/test_tasks/test_opm_operations.py [hunk 16]
```diff
@@ -373,3 +373,41 @@ def test_opm_registry_rm(mock_run_cmd):
     opm_args = mock_run_cmd.call_args[0][0]
     assert opm_args[:3] == ['opm', 'registry', 'rm']
     assert ','.join(packages) in opm_args
+
+
+@pytest.mark.parametrize(
+    'from_index, is_fbc', [('some-fbc-index:latest', True), ('some-sqllite-index:latest', False)]
```
Comment (pbortlov) [inline tests/test_workers/test_tasks/test_opm_operations.py:379]: Could we parametrize approach with `from_index`, `is_fbc` and `result` option? This way we could avoid using If statements in tests.

Comment (pbortlov) [reply tests/test_workers/test_tasks/test_opm_operations.py:379]: or add a path to the database as one parameter. This could make it easier

**Review by pbortlov — CHANGES_REQUESTED** [review]

File: iib/web/api_v1.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (chandwanitulsi) [reply iib/web/api_v1.py:898]: @lipoja @pbortlov  Asking for your preference again here before resolving this comment

File: iib/workers/tasks/build_create_empty_index.py (hunk 13 above)
Comment (chandwanitulsi) [reply iib/workers/tasks/build_create_empty_index.py:92]: _get_present_operators() function hides the implementation details well, and  takes care of getting operators even index image is FBC.

More information: the function calls is as follows:
_get_present_operators -> grpcurl_get_db_data -> opm_serve_from_index()
opm_serve_from_index() function determines if the index image is fbc or sqlite based and accordingly calls opm_serve or opm_registry_serve() respectively

Comment (pbortlov) [reply iib/workers/tasks/build_create_empty_index.py:92]: thanks for explaining

File: tests/test_workers/test_tasks/test_build_create_empty_index.py (hunk 15 above)
Comment (chandwanitulsi) [reply tests/test_workers/test_tasks/test_build_create_empty_index.py:127]: this is a test to ensure IIBerror is raised when fbc=False, but from_index provided is FBC index image. Let me change the function name to make it more intuitive

File: iib/web/api_v1.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (pbortlov) [reply iib/web/api_v1.py:898]: Well, If check the names of other variables it doesn't make sense to have `create_fbc` and if the name of the function in the worker is **bool** `is_fbc`

File: iib/workers/tasks/build_create_empty_index.py [hunk 17]
```diff
@@ -86,16 +87,30 @@ def handle_create_empty_index_request(
...
+            log.debug('Creating empty FBC index image from %s', from_index)
+            opm_create_empty_fbc(
+                request_id=request_id,
+                temp_dir=temp_dir,
+                from_index_resolved=from_index_resolved,
+                from_index=from_index,
+                binary_image=binary_image,
+                operators=operators,
+            )
+        else:
```
Comment (pbortlov) [inline iib/workers/tasks/build_create_empty_index.py:106]: what if the from_index is FBC? Should we check it?

Comment (chandwanitulsi) [reply iib/workers/tasks/build_create_empty_index.py:106]: opm_create_empty_fbc already checks if the from_index is fbc or not.
also there is one more case, when fbc parameter is false, but from_index given is an fbc image, that case is checked in line 107

File: iib/web/api_v1.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (pbortlov) [reply iib/web/api_v1.py:898]: I am fine with fbc it is better

File: tests/test_workers/test_tasks/test_build_create_empty_index.py (hunk 15 above)
Comment (pbortlov) [reply tests/test_workers/test_tasks/test_build_create_empty_index.py:127]: thanks

Comment (chandwanitulsi) [reply tests/test_workers/test_tasks/test_build_create_empty_index.py:129]: Added a comment to describe the test as well

File: iib/web/api_v1.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (lipoja) [reply iib/web/api_v1.py:898]: @chandwanitulsi @pbortlov Sure I am fine with anything. 

I was just thinking about the curl command and its payload. Not from IIB point of view but from user perspective. What fbc would mean if you see that parameter in curl command. That is why I was thinking about something like `create_fbc` or `output_fbc`.

[the remark this answers is not in the document]
Comment (lipoja) [reply iib/web/api_v1.py:898]: If you want to keep it feel free to resolve this.

File: iib/web/static/api_v1.yaml [hunk 18]
```diff
@@ -1143,6 +1143,9 @@ components:
           items:
             type: string
           example: ["v4.5-10-08-2021"]
+        fbc:
```
Comment (lipoja) [inline iib/web/static/api_v1.yaml:1146]: The other properties are having `example` section. Should we add it here as well?

File: iib/workers/tasks/build_create_empty_index.py [hunk 19]
```diff
@@ -58,21 +59,21 @@ def _get_present_operators(from_index, base_dir):
...
 def handle_create_empty_index_request(
-    from_index, request_id, binary_image=None, labels=None, binary_image_config=None
+    from_index, request_id, fbc=False, binary_image=None, labels=None, binary_image_config=None
 ):
     """Coordinate the the work needed to create the index image with labels.
 
     :param str from_index: the pull specification of the container image containing the index that
         the index image build will be based from.
     :param int request_id: the ID of the IIB build request
+    :param bool fbc: specifies whether a File Based Catalog index image should be created
```
Comment (lipoja) [inline iib/workers/tasks/build_create_empty_index.py:69]: I am not native speaker so I am not sure what is right `File Based Catalog` or `File-Based Catalog` with `-` sign?

File: iib/workers/tasks/build_create_empty_index.py [hunk 20]
```diff
@@ -86,16 +87,30 @@ def handle_create_empty_index_request(
...
+                request_id=request_id,
+                temp_dir=temp_dir,
+                from_index_resolved=from_index_resolved,
+                from_index=from_index,
+                binary_image=binary_image,
+                operators=operators,
+            )
+        else:
+            if is_image_fbc(from_index_resolved):
+                err_msg = 'File-Based catalog image type is not supported when fbc is None/False'
```
Comment (lipoja) [inline iib/workers/tasks/build_create_empty_index.py:108]: What about message: `Cannot create SQLite index image from File-Based Catalog index image.`
Or something similar?

File: tests/test_workers/test_tasks/test_build_create_empty_index.py [hunk 21]
```diff
@@ -120,12 +121,99 @@ def test_handle_create_empty_index_request(
 @mock.patch('iib.workers.tasks.build_create_empty_index._cleanup')
 @mock.patch('iib.workers.tasks.build_create_empty_index.is_image_fbc')
 def test_handle_create_empty_index_request_raises(mock_iifbc, mock_c):
+    # Ensure that IIb raises error when fbc paramater is false
```
Comment (lipoja) [inline tests/test_workers/test_tasks/test_build_create_empty_index.py:124]: Typo: IIb -> IIB

File: tests/test_workers/test_tasks/test_build_create_empty_index.py [hunk 22]
```diff
@@ -120,12 +121,99 @@ def test_handle_create_empty_index_request(
...
+    # but from_index provided is FBC
     mock_iifbc.return_value = True
     with pytest.raises(IIBError):
```
Comment (lipoja) [inline tests/test_workers/test_tasks/test_build_create_empty_index.py:127]: May I suggest to modify this and add `match` parameter to check if expected error message was raised `with pytest.raises(IIBError, match=error_msg):`

File: iib/workers/tasks/build_create_empty_index.py (hunk 17 above)
Comment (JAVGan) [reply iib/workers/tasks/build_create_empty_index.py:106]: I don't understand why do we have the boolean flag `fbc` then, can you please elaborate?

If `opm_create_empty_fbc` can know whether the image is File-Based Catalog or not with just the information of `from_index` why do we need an additional flag here?

Isn't possible to use a similar logic to determine whether the index image is fbc or not using just `from_index` here as well? Then we wouldn't need even the check from line 107 IMO, but it's just a question :slightly_smiling_face:

File: iib/web/api_v1.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (chandwanitulsi) [reply iib/web/api_v1.py]: I changed it to output_fbc

File: tests/test_workers/test_tasks/test_opm_operations.py (hunk 16 above)
Comment (chandwanitulsi) [reply tests/test_workers/test_tasks/test_opm_operations.py]: Since its an assertion check, I let it be in if else

File: iib/web/static/api_v1.yaml (hunk 18 above)
Comment (chandwanitulsi) [reply iib/web/static/api_v1.yaml]: Added example and default

File: iib/workers/tasks/build_create_empty_index.py (hunk 19 above)
Comment (chandwanitulsi) [reply iib/workers/tasks/build_create_empty_index.py]: Changed it already

File: iib/workers/tasks/build_create_empty_index.py (hunk 20 above)
Comment (chandwanitulsi) [reply iib/workers/tasks/build_create_empty_index.py]: Done. changed

File: tests/test_workers/test_tasks/test_build_create_empty_index.py (hunk 21 above)
Comment (chandwanitulsi) [reply tests/test_workers/test_tasks/test_build_create_empty_index.py]: Changed

File: tests/test_workers/test_tasks/test_build_create_empty_index.py (hunk 22 above)
Comment (chandwanitulsi) [reply tests/test_workers/test_tasks/test_build_create_empty_index.py]: Added, Thanks for suggestion, ended up improving the test and optimizing the code :)

File: iib/workers/tasks/build_create_empty_index.py (hunk 17 above)
Comment (lipoja) [reply iib/workers/tasks/build_create_empty_index.py:118]: We want to specifically say that we want to create FBC image - by setting the `fbc` flag. 
Otherwise you are not able to create empty FBC index image. Because at this point you do not have any FBC image created. 

So the paths for create empty index are:
1. Input SQLite -> output SQLite
2. Input SQLite -> output FBC (flag must be set)
3. Input FBC -> output FBC  (flag must be set)

There is no backward compatibility and we can not generate SQLite from FBC image. Therefore IIB has to fail when FBC is on input and fbc flag is not set.

File: tests/test_workers/test_tasks/test_build_create_empty_index.py [hunk 23]
```diff
@@ -119,13 +120,116 @@ def test_handle_create_empty_index_request(
...
+    mock_prfb.return_value = {
+        'arches': arches,
+        'binary_image': binary_image,
+        'binary_image_resolved': binary_image_resolved,
+        'from_index_resolved': from_index_resolved,
+        'ocp_version': 'v4.5',
+        'distribution_scope': 'prod',
+    }
+    with pytest.raises(
+        IIBError, match=('Cannot create SQLite index image from File-Based' ' Catalog index image')
```
Comment (lipoja) [inline tests/test_workers/test_tasks/test_build_create_empty_index.py:141]: I guess black formatted the text. Can you merge it to one string, please?

**Comment by lipoja** [comment]
@chandwanitulsi Just the small one to improve reading the code. Otherwise it looks OK. Ping me when it is ready for final review.
Thanks!

**Review by JAVGan — APPROVED** [review]
LGTM: `output_fbc` is more clear for the flag, thanks a lot!

Please make sure to rebase it :slightly_smiling_face:

**Review by JAVGan — APPROVED** [review]

**Review by lipoja — APPROVED** [review]
